src/core/net/netif.hpp

```diff
@@ -164,8 +200,125 @@ class Netif : public InstanceLocator, public LinkedListEntry<Netif>, private Non
 {
     friend class Ip6;
     friend class Address;
+    class ExternalMulticastAddressIteratorBuilder;
 
 public:
+    /**
+     * This class represents an iterator for iterating external multicast addresses in a Netif instance.
+     *
+     */
+    class ExternalMulticastAddressIterator
+    {
+        friend class ExternalMulticastAddressIteratorBuilder;
+
+    public:
+        /**
+         * This constructor initializes an `ExternalMulticastAddressIterator` instance to start from the first external
+         * multicast address.
+         *
+         * @param[in] aNetif  A reference to the Netif instance.
+         *
+         */
+        explicit ExternalMulticastAddressIterator(Netif &aNetif)
+            : mNetif(aNetif)
+        {
+            AdvanceFrom(const_cast<NetifMulticastAddress *>(mNetif.GetMulticastAddresses()));
+        }
+
+        /**
+         * This method overloads `++` operator (pre-increment) to advance the iterator.
+         *
+         * The iterator is moved to point to the next entry.  If there are no more entries matching the iterator becomes
+         * empty.
+         *
+         */
+        void operator++(void) { AdvanceFrom(mCurrent->GetNext()); }
```

Comment:
AdvanceFrom(const_cast<NetifMulticastAddress *>(mNetif.GetMulticastAddresses()));

Reply:
I intentionally didn't check it. I think we should never call Advance when the iterator is empty because it is a logical error. We will never have any issue if we just use it in c++11 loops.

Reply:
The document of [std::advance](https://en.cppreference.com/w/cpp/iterator/advance) defines increasing the `past-the-end` iterator to be undefined. 
```
The behavior is undefined if the specified sequence of increments or decrements would require that 
a non-incrementable iterator (such as the past-the-end iterator) is incremented, ...
```

So I think it's not required to check for `mCurrent == nullptr` so that we have a more concise implementation.

Reply:
SGTM.